src/asgen/backends/alpinelinux/apkpkgindex.d

```diff
@@ -0,0 +1,202 @@
+import std.file : dirEntries, exists, SpanMode;
+import std.format : format;
+import std.path : baseName, buildPath;
+import std.string : splitLines, startsWith, strip;
+import std.utf : UTFException, validate;
+
+import asgen.logging;
+import asgen.zarchive;
+import asgen.utils : escapeXml;
+import asgen.backends.interfaces;
+import asgen.backends.alpinelinux.apkpkg;
+
+final class AlpinePackageIndex : PackageIndex
+{
+
+private:
+    string rootDir;
+    Package[][string] pkgCache;
+
+public:
+
+    this(string dir)
+    {
+        enforce(exists(dir), format("Directory '%s' does not exist.", dir));
+        this.rootDir = dir;
+    }
+
+    override void release()
+    {
+        pkgCache = null;
+    }
+
+    private void setPkgDescription(AlpinePackage pkg, string pkgDesc)
+    {
+        if (pkgDesc is null)
+        {
+            return;
+        }
+
+        auto desc = "<p>%s</p>".format(pkgDesc.escapeXml);
+        pkg.setDescription(desc, "C");
+    }
+
+    private Package[] loadPackages(string suite, string section, string arch)
+    {
+        auto apkRootPath = buildPath(rootDir, suite, section, arch);
+        ArchiveDecompressor ad;
+        AlpinePackage[string] pkgsMap;
+
+        foreach (packageArchivePath; dirEntries(apkRootPath, SpanMode.shallow).filter!(
+                f => f.name.endsWith(".apk")))
+        {
+            auto fileName = packageArchivePath.baseName();
+            AlpinePackage pkg;
+            if (fileName in pkgsMap)
+            {
+                pkg = pkgsMap[fileName];
+            }
+            else
+            {
+                pkg = new AlpinePackage();
+                pkgsMap[fileName] = pkg;
+            }
+
+            ad.open(packageArchivePath);
+            auto pkgInfoData = cast(string) ad.readData(".PKGINFO");
+
+            try
+            {
+                validate(pkgInfoData);
+            }
+            catch (UTFException e)
+            {
+                logError("PKGINFO file in archive %s contained invalid UTF-8, skipping!",
+                        packageArchivePath);
+            }
+
+            pkg.filename = packageArchivePath;
+            auto lines = pkgInfoData.splitLines();
+            for (auto i = 0; i < lines.length; i++)
+            {
+                /* Skip comments */
+                if (lines[i].startsWith("#"))
+                {
+                    continue;
+                }
+
+                auto line = lines[i].strip().split("#")[0];
+
+                /*
+                    When the next line doesn't contain a "=" it's a continuation
+                    of the current line
+                */
+                int j = i + 1;
+                while (lines.length > j && !lines[j].canFind("=") && !lines[j].startsWith("#"))
+                {
+                    line ~= " " ~ lines[j].strip().split("#")[0];
+                    /* Skip the next line since it was a continuation of the previous one */
+                    i++;
```

Comment:
This is ugly - if `i` went over the array limit and is used anywhere after this point, this will crash.
One possible fix is not not look ahead at the next line, but instead remember the current value until reading the next line and to only finalize the value when that line doesn't contain an equal sign. That way, you can save a loop and get rid of all index variables by using a foreach loop for `lines`.

Reply:
I've switched this for what you suggested - I initially though this would be way uglier which is why I did this first, but I think it turned out OK.